FirebaseRemoteConfig/SwiftNew/ConfigRealtime.swift

```diff
@@ -0,0 +1,665 @@
+                                 code: RemoteConfigError.internalError.rawValue,
+                                 userInfo: userInfo)
+            )
+          } else if let identifier = identifier {
+            RCLog.info("I-RCN000022", "Success to get iid : \(identifier)")
+            self.reportCompletion(onHandler: completionHandler,
+                                  withStatus: .noFetchYet,
+                                  withError: nil)
+          }
+        }
+      }
+    }
+  }
+
+  @objc public
+  func createRequestBody(completion: @escaping (Data) -> Void) {
+    refreshInstallationsToken { status, error in
+      if self.settings.configInstallationsIdentifier.isEmpty {
+        RCLog.debug(
+          "I-RCN000013",
+          "Installation token retrieval failed. Realtime connection will not include " +
+            "valid installations token."
+        )
+      }
+      var request = self.request
+      request.setValue(self.settings.configInstallationsToken,
+                       forHTTPHeaderField: installationsAuthTokenHeaderName)
+      if let etag = self.settings.lastETag {
+        request.setValue(etag, forHTTPHeaderField: ifNoneMatchETagHeaderName)
+      }
+
+      let postBody = """
+      {
+      project:'\(self.options.gcmSenderID)',
+      namespace:'\(Utils.namespaceOnly(self.namespace))',
+      lastKnownVersionNumber:'\(self.configFetch.templateVersionNumber)',
+      appId:'\(self.options.googleAppID)',
+      sdkVersion:'\(RemoteConfig.sdkVersion())',
+      appInstanceId:'\(self.settings.configInstallationsIdentifier)'
+      }
+      """
+      do {
+        if let postData = postBody.data(using: .utf8) {
+          let compressedData = try NSData.gul_data(byGzippingData: postData)
+          completion(compressedData)
+        } else {
+          RCLog.error("I-RCN000090", "Error creating fetch body for realtime")
+          completion(Data())
+        }
+      } catch {
+        RCLog.error("I-RCN000091", "Error compressing fetch body for realtime \(error)")
+        completion(Data())
+      }
+    }
+  }
+
+  // MARK: - Retry Helpers
+
+  func canMakeConnection() -> Bool {
+    let noRunningConnection = dataTask == nil || dataTask?.state != .running
+    return noRunningConnection && listeners.count == 0 && !isInBackground && !isRealtimeDisabled
+  }
+
+  func retryHTTPConnection() {
+    realtimeLockQueue.async {
+      guard self.remainingRetryCount > 0 else {
+        let error = NSError(domain: RemoteConfigUpdateErrorDomain,
+                            code: RemoteConfigUpdateError.streamError.rawValue,
+                            userInfo: [
+                              NSLocalizedDescriptionKey: "Unable to connect to the server. Check your connection and try again.",
+                            ])
+        RCLog.error("I-RCN000014", "Cannot establish connection. Error: \(error)")
+        self.propagateErrors(error)
+        return
+      }
+      if self.canMakeConnection() {
```

Comment:
optional, but consider moving this up to the guard condition:

guard self.canMakeConnection(), self.remainingRetryCount > 0  else { ... }

Reply:
Not sure about that, since it could introduce more errors when there are no listeners?